### PR TITLE
Added short-hash support to downloads

### LINKED ISSUE #211: Missing support for short-hash archives download
While short-hash are recognised by the commit browser, for example:

 *  https://try.gogs.io/strk/trac2gogs/commit/addbbc52

The same short reference doesn't work at fetching a tarball:

 * https://try.gogs.io/strk/trac2gogs/archive/addbbc52.tar.gz

While full hash does work:

 * https://try.gogs.io/strk/trac2gogs/archive/addbbc524c52f271c2542009bc947c8200b588a9.tar.gz

It would be nice if hashes recognition was consistent.

See how the first one was fixed in https://github.com/gogits/gogs/issues/3321

Ticket migrated from https://github.com/gogits/gogs/issues/3834

### PR DESCRIPTION
**Pull request by Bwko**
Fixes #211

### CODE REVIEW

**Comment by strk** [comment]
LGTM - this should go in 1.0.0 as bugfix

File: routers/repo/repo.go [hunk 1]
```diff
@@ -311,7 +311,7 @@ func Download(ctx *context.Context) {
 			ctx.Handle(500, "GetTagCommit", err)
 			return
 		}
-	} else if len(refName) == 40 {
+	} else if len(refName) > 7 && len(refName) <= 40 {
```
Comment (strk) [inline routers/repo/repo.go]: btw, should that be >= 7 instead ?

Comment (tboerger) [inline routers/repo/repo.go]: It should be == 7 or == 40, beside that I'm not sure if this is the only place that needs to be changed

**Comment by strk** [comment]
Adding an automated testcase would be great too

**Review by tboerger — CHANGES_REQUESTED** [review]

**Comment by Bwko** [comment]
Currently only a short hash of 8 characters work. But it's not mandatory for a short hash to be x characters long. See http://stackoverflow.com/a/21015031
Maybe we could set it from >=4 (minimum short hash is 4 characters) to <= 40?

**Comment by ghost** [comment]
LGTM

File: routers/repo/repo.go (hunk 1 above)
Comment (bkcsoft) [reply routers/repo/repo.go]: nope, should be `7 <= len(refName) <= 40` just like it is, otherwise `f00badf00d.tar.gz` wouldn't work 🙂

Comment (tboerger) [reply routers/repo/repo.go]: So how have it worked before with == 40?

**Comment by strk** [comment]
Before it didn't work for anything != 40, follow the reference
to the bug (filed by myself, btw).

The new code will still not support a 7-chars long refName
due to `len(refName) > 7` -- should read `len(refName) >= 7`,
or whatever min lenght you want to support.

Question: what happens if a colliding short hash is used ?

**Comment by Bwko** [comment]
That's why I suggested this:

> Maybe we could set it from >=4 (minimum short hash is 4 characters) to <= 40?

Question: what happens if a colliding short hash is used ?

Git will produce an ambiguous error:

> error: short SHA1 0001 is ambiguous.

**Comment by strk** [comment]
On Thu, Nov 24, 2016 at 12:10:32AM -0800, Bwko wrote:
> That's why I suggested this:
> 
> > Maybe we could set it from >=4 (minimum short hash is 4 characters) to <= 40?
> 
> Question: what happens if a colliding short hash is used ?
> 
> Git will produce an ambiguous error:
> 
> > error: short SHA1 0001 is ambiguous.

Is Gitea code ready to handle that or would it choke ?

**Comment by thibaultmeyer** [comment]
Github allow usage of 7 length commit hashes. No collision encountered on big repo (eg: Linux kernel have 700k commits).

By default, GITEA use 10 length short hash on all pages. Maybe we can change routes to accept at least 10 length hash

**Comment by Bwko** [comment]
I think it'll throw a 404 error. When using 7 characters the change of a collision is 1 in 250 million

**Comment by strk** [comment]
On Thu, Nov 24, 2016 at 12:20:43AM -0800, Bwko wrote:
> I think it'll throw a 404 error.

Could you please check ?
Accept as low as 2 characters and try fetching a tarball with those.

> When using 7 characters the change of a collision is 1 in 250 million

Is there any reason for not accepting a few as 4 chars, from an API
point of view ? This is about *accepting*, not *providing*, so we
should be as liberal as possible, for the robustness principle.

I'm fine with starting as low as 7, in any case (but this PR starts
from 8 as far as I understand the code, so should still be changed)

**Comment by Bwko** [comment]
@strk I've lowered the minimum length to 4. I've just tested this and collisions result in a 404 error

**Review by tboerger — APPROVED** [review]

**Comment by tboerger** [comment]
LGTM

**Comment by strk** [comment]
LGTM, please merge @go-gitea/owners

**Pull request merged**